staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -41,15 +43,21 @@ import (
 	"k8s.io/utils/trace"
 )
 
+const defaultExpectedTypeName = "<unspecified>"
+
 // Reflector watches a specified resource and causes all changes to be reflected in the given store.
 type Reflector struct {
 	// name identifies this reflector. By default it will be a file:line if possible.
 	name string
 	// metrics tracks basic metric information about the reflector
 	metrics *reflectorMetrics
 
-	// The type of object we expect to place in the store.
-	expectedType reflect.Type
```

Comment:
would it be simpler to keep the expectedType aspect as-is, and add an optional `expectedGVK *schema.GroupVersionKind` that we populate only in unstructured cases?

Reply:
That was effectively what I originally proposed (at least in terms of variables) - add `expectedGVK` and `expectedTypeName` (to simplify logging of either either `expectedType.String()` or `expectedGVK.String()`. The renames to `expectedNativeType` and `expectedUnstructuredGVK` were requested by @sttts:

https://github.com/kubernetes/kubernetes/pull/82388#discussion_r321451910
https://github.com/kubernetes/kubernetes/pull/82388#discussion_r321450959

Reply:
Apologies, I think I misinterpreted your intent. Would you prefer to leave the check of the reflected type as-is (always check if non-nil, even if unstructured), and add the check for gvk if provided by unstructured?

Regarding the logging, is it acceptable to log gvk instead of reflected type name if gvk were provided? Without this change, the log output of multiple informers with an unstructured expected type is undifferentiated.

Reply:
> Would you prefer to leave the check of the reflected type as-is (always check if non-nil, even if unstructured), and add the check for gvk if provided by unstructured?

Yes, exactly.

> Regarding the logging, is it acceptable to log gvk instead of reflected type name if gvk were provided? Without this change, the log output of multiple informers with an unstructured expected type is undifferentiated.

Yes, improving the logging is fine.

Reply:
I've updated the check for expected type to be effectively the same as it was previously. Would you prefer that I go further and revert to using `expectedType` instead of `expectedNativeType`?

Reply:
> Would you prefer that I go further and revert to using expectedType instead of expectedNativeType?

I think so, since we check it in all cases

Reply:
Done.